src/gui/attributetable/qgsattributetablemodel.cpp

```diff
@@ -340,10 +351,70 @@ void QgsAttributeTableModel::loadAttributes()
   }
 }
 
+
+void QgsAttributeTableModel::loadLayerFinished()
+{
+  QgsDebugMsg( "loadLayerFinished" );
+  endResetModel();
+  emit loadFinished();
+}
+
+
+void QgsAttributeTableModel::featuresReady( QgsFeatureList features, int loadedCount )
+{
+  QgsDebugMsg( "featuresReady" );
+  bool cancel = false;
+  emit loadProgress( loadedCount, cancel );
+  if ( cancel )
+  {
+    if ( mLoadWorker && mLoadWorker->isRunning() )
+    {
+      mLoadWorker->stopJob();
+    }
+    return;
+  }
+  QgsFeatureId fid;
+  int n = mRowIdMap.size();
+  int m = n + features.count() - 1;
+  beginInsertRows( QModelIndex(), n, m );
+  Q_FOREACH ( const QgsFeature &f, features )
+  {
+    mFeat = f;
+    fid = f.id();
+    // Is this check really necessary?
+    if ( mFeatureRequest.acceptFeature( mFeat ) )
+    {
+      mFieldCache[ fid ] = mFeat.attribute( mCachedField );
+      mIdRowMap.insert( fid, n );
+      mRowIdMap.insert( n, fid );
+    }
+    n++;
+  }
+  endInsertRows();
+  reload( index( rowCount() - 1, 0 ), index( rowCount() + features.count() - 1, columnCount() ) );
+}
+
+
 void QgsAttributeTableModel::loadLayer()
 {
   QgsDebugMsg( "entered." );
 
+  // Stop old thread
+  if ( mLoadWorker && mLoadWorker->isRunning() )
+  {
+    mLoadWorker->stopJob();
+  }
+
+  QgsFeatureIterator features = mLayerCache->getFeatures( mFeatureRequest );
+
+  // Pass the total number of features as a maximum for the progress bar
+  long fc = mLayerCache->layer()->featureCount();
+  emit loadStarted( fc );
+
+  // Ensure signals are delivered
+  qApp->processEvents();
```

Comment:
I think there are other occurrences in the attribute table which were previously necessary and no longer are with this patch.

It is dangerous because it executes code in a context which it was normally not designed to be run in.
E.g. a user action could lead to deletion of an object in the call stack.